extension/src/pipeline/index.ts

```diff
@@ -1,34 +1,195 @@
+import { quickPickOne, quickPickOneOrInput } from '../vscode/quickPick'
+import { pickFile } from '../vscode/resourcePicker'
+
+export enum ScriptCommand {
+  JUPYTER = 'jupyter nbconvert --to notebook --inplace --execute',
+  PYTHON = 'python'
+}
+
+const getScriptCommand = (script: string) => {
+  switch (getFileExtension(script)) {
+    case '.py':
+      return ScriptCommand.PYTHON
+    case '.ipynb':
+      return ScriptCommand.JUPYTER
+    default:
+      return ''
+  }
+}
 
 export class Pipeline extends DeferredDisposable {
+  public onDidUpdate: Event<void>
+
+  private updated: EventEmitter<void>
+
   private readonly dvcRoot: string
   private readonly data: PipelineData
+  private readonly model: PipelineModel
 
-  constructor(dvcRoot: string, internalCommands: InternalCommands) {
+  constructor(
+    dvcRoot: string,
+    internalCommands: InternalCommands,
+    data?: PipelineData
+  ) {
     super()
     this.dvcRoot = dvcRoot
-    this.data = this.dispose.track(new PipelineData(dvcRoot, internalCommands))
+    this.data = this.dispose.track(
+      data || new PipelineData(dvcRoot, internalCommands)
+    )
+    this.model = this.dispose.track(new PipelineModel())
+    this.updated = this.dispose.track(new EventEmitter<void>())
+    this.onDidUpdate = this.updated.event
 
     void this.initialize()
   }
 
+  public hasPipeline() {
+    return this.model.hasPipeline()
+  }
+
+  public async getCwd() {
+    await this.checkOrAddPipeline()
+
+    const pipelines = this.model.getPipelines()
+    if (!pipelines?.size) {
+      return
+    }
+    if (pipelines.has(this.dvcRoot)) {
+      return this.dvcRoot
+    }
+    if (pipelines.size === 1) {
+      return [...pipelines][0]
+    }
+
+    return quickPickOne(
+      [...pipelines],
+      'Select a Pipeline to Run Command Against'
+    )
+  }
+
+  public checkOrAddPipeline() {
+    if (this.model.hasPipeline()) {
+      return
+    }
+    return this.addPipeline()
+  }
+
   public forceRerender() {
     return appendFileSync(join(this.dvcRoot, TEMP_DAG_FILE), '\n')
   }
 
   private async initialize() {
     this.dispose.track(
-      this.data.onDidUpdate(data =>
-        writeFileSync(join(this.dvcRoot, TEMP_DAG_FILE), data)
-      )
+      this.data.onDidUpdate(({ dag, stages }) => {
+        this.writeDag(dag)
+        const hasPipeline = this.model.hasPipeline()
+        this.model.transformAndSet(stages)
+        if (hasPipeline !== this.model.hasPipeline()) {
+          this.updated.fire()
+        }
+      })
     )
+    void this.data.managedUpdate()
 
     await this.data.isReady()
     return this.deferred.resolve()
   }
+
+  private async addPipeline() {
+    const stageName = await this.askForStageName()
+    if (!stageName) {
+      return
+    }
+
+    const { trainingScript, command, enteredManually } =
+      await this.askForTrainingScript()
+    if (!trainingScript) {
+      return
+    }
+
+    const dataUpdated = new Promise(resolve => {
```

Comment:
[F] This promise is added so that if a user tries to run an experiment and has to add a stage the experiment will start running after the new stage has been added.